### PR TITLE
explore some parsing helpers that could act as the other side of the StreamId.gen coin

### PR DESCRIPTION
**Pull request by nordfjord**

### CODE REVIEW

File: src/FsCodec/StreamName.fs [hunk 1]
```diff
@@ -128,3 +128,20 @@ module StreamName =
     /// Throws <c>InvalidArgumentException</c> if the stream name is not well-formed</summary>
     /// <remarks>Inverse of <c>create</c></remarks>
     let (|CategoryAndIds|) : StreamName -> struct (string * string[]) = splitCategoryAndIds
+
+    module Parse =
+        let parse category f (name: StreamName) =
```
Comment (bartelink) [inline src/FsCodec/StreamName.fs:133]: blank line, inside module and at EOF
can you do a 4th one too - streamId ones have 4
also why no pattern match to pull the ids out of the array and validate len ? (can use `CategoryAndIds` pattern)

File: tests/FsCodec.Tests/StreamNameTests.fs [hunk 2]
```diff
@@ -35,3 +35,39 @@ let [<Fact>] ``StreamName parse throws given 0 separators`` () =
...
+    | StreamId2 _ -> failwith "Unexpected match"
+    | _ -> ()
+
+let [<Fact>] ``The single id matcher doesn't care about underscored`` () =
+    match StreamName.create "Cat" "a-b_c-d_e-f" with
+    | StreamId1 x -> test <@ x = "a-b_c-d_e-f" @>
+    | _ -> failwith "Unexpected match"
+
+
+let [<Fact>] ``It works with Guid's`` () =
```
Comment (bartelink) [inline tests/FsCodec.Tests/StreamNameTests.fs:68]: there's no ' in Guids (or id's!) ;)

File: src/FsCodec/StreamId.fs [hunk 3]
```diff
@@ -0,0 +1,37 @@
+namespace FsCodec 
+
+open FSharp.UMX
+type private String = System.String
+
+/// Represents the second half of a canonical StreamName, i.e., the streamId in "{categoryName}-{streamId}"
+type StreamId = string<streamId>
+and [<Measure>] streamId
+/// Low-level helpers for composing and rendering StreamId values; prefer the ones in the Equinox namespace
```
Comment (bartelink) [inline src/FsCodec/StreamId.fs:9]: these guys belong in a `module Internal`

File: src/FsCodec/StreamId.fs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    /// Combines streamId fragments. Throws if any of the fragments embed a `_`, are `null`, or are empty
+    let ofFragments (fragments: string[]): StreamId =
+        fragments |> Array.iter StreamName.Internal.validateElement 
+        String.Join(separator, fragments) |> ofRaw
+    /// Render as a string for external use
+    let toString: StreamId -> string = UMX.untag
+    /// Render as a canonical "{categoryName}-{streamId}" StreamName. Throws if the categoryName embeds `-` chars.
+    let renderStreamName categoryName (x: StreamId): StreamName = toString x |> StreamName.create categoryName
+
+    /// Generate a StreamId from a single application-level id, given a rendering function that maps to a non empty fragment without embedded `_` chars
```
Comment (bartelink) [inline src/FsCodec/StreamId.fs:29]: dont forget to bring back the `type` - these are ugly to all from C# (and there is a sample that uses it)

File: src/FsCodec/StreamId.fs [hunk 5]
```diff
@@ -0,0 +1,37 @@
+namespace FsCodec
```
Comment (bartelink) [inline src/FsCodec/StreamId.fs:1]: module FsCodec.StreamId?

File: src/FsCodec/StreamName.fs [hunk 6]
```diff
@@ -129,19 +129,28 @@ module StreamName =
...
-            if streamCategory = category && ids.Length = 2
-            then ValueSome struct(f ids[0], g ids[1])
-            else ValueNone
-        let parse3 category f g h (name: StreamName) =
-            let struct(streamCategory, ids) = splitCategoryAndIds name
-            if streamCategory = category && ids.Length = 3
-            then ValueSome struct(f ids[0], g ids[1], h ids[2])
-            else ValueNone
+    /// Extracts the application-level id of a StreamName
+    let dec category f (name: StreamName) = function
```
Comment (bartelink) [inline src/FsCodec/StreamName.fs:133]: should split the streamId and throw if >1 item

**Comment by nordfjord** [comment]
Superseded by #100

**Pull request closed without merging**